### PR TITLE
[dataset] add `ValidateTlvs()`, helper methods, and unit test

### PR DESCRIPTION
**Pull request by abtink**
This commit enhances dataset handling:

- Adds `ValidateTlvs()` which parses and validates all known TLVs within the `Dataset` and checks for any duplicates.
- Introduces a set of `SetFrom()` methods for constructing a `Dataset` from various types of inputs, e.g., other `Dataset`, TLV sequences, `DatasetInfo` structures, bytes read from `Message`. Now they consistently clear the `Dataset` before setting it.
- Adds `WriteTlvsFrom()` to update a `Dataset` replacing/appending a set of TLVs.
- Adds `AppendTlvsFrom()` to append an already encoded sequence of TLVs without validating/checking the format.
- Renames `Get/SetSize()` to `Get/SetLength()` for consistency.
- Adds `test_dataset.cpp` unit test for basic `Dataset` validation.

### CODE REVIEW

File: src/core/meshcop/dataset.hpp [hunk 1]
```diff
@@ -496,48 +562,57 @@ class Dataset
...
-    Error ReadFromMessage(const Message &aMessage, uint16_t aOffset, uint16_t aLength);
+    void SetFrom(const Info &aDatasetInfo);
 
     /**
-     * Sets the Dataset using an existing Dataset.
+     * Sets the Dataset from a given sequence of TLVs.
      *
-     * If this Dataset is an Active Dataset, any Pending Timestamp and Delay Timer TLVs will be omitted in the copy
-     * from @p aDataset.
+     * @param[in]  aDataset  The input Dataset as `Tlvs`.
```
Comment (Irving-cl) [inline src/core/meshcop/dataset.hpp:583]:
```suggestion
     * @param[in]  aTlvs  The input Dataset as `Tlvs`.
```

File: src/core/meshcop/dataset_local.hpp [hunk 2]
```diff
@@ -167,6 +167,7 @@ class DatasetLocal : public InstanceLocator
      * @param[in]  aDatasetTlvs  The Dataset to save as `Dataset::Tlvs`.
      *
      * @retval kErrorNone             Successfully saved the dataset.
+     * @retval kErrorInvalidArgs      The dataset TLVs is invalid, its length is longer than `Dataset::kMaxSize`.
```
Comment (Irving-cl) [inline src/core/meshcop/dataset_local.hpp:170]:
```suggestion
     * @retval kErrorInvalidArgs      The dataset TLVs is invalid, its length is longer than `Dataset::kMaxLength`.
```

**Review by Irving-cl — APPROVED** [review]
LGTM 👍 

A few minor comments.

File: src/core/meshcop/dataset_local.hpp (hunk 2 above)
Comment (abtink) [reply src/core/meshcop/dataset_local.hpp:170]: Thanks.

**Review by jwhui — APPROVED** [review]

**Pull request merged**